pkg/plugins/policies/core/rules/resourcerules.go

```diff
@@ -0,0 +1,172 @@
+package rules
+
+import (
+	"slices"
+
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	meshextenralservice_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshexternalservice/api/v1alpha1"
+	meshservice_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshservice/api/v1alpha1"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	meshhttproute_api "github.com/kumahq/kuma/pkg/plugins/policies/meshhttproute/api/v1alpha1"
+)
+
+type ResourceRule struct {
+	Resource core_model.ResourceMeta
+	Conf     interface{}
+	Origin   []core_model.ResourceMeta
+}
+
+type UniqueResourceKey string
+
+type ResourceRules map[UniqueResourceKey]ResourceRule
+
+func (rr ResourceRules) Compute(r core_model.Resource, l ResourceLister) *ResourceRule {
+	key := uniqueKey(r)
+	if rule, ok := rr[key]; ok {
+		return &rule
+	}
+
+	findMesh := func(meshName string) core_model.Resource {
+		meshes := l.ListOrEmpty(core_mesh.MeshType).GetItems()
+		if i := slices.IndexFunc(meshes, func(resource core_model.Resource) bool {
+			return resource.GetMeta().GetName() == meshName
+		}); i >= 0 {
+			return meshes[i]
+		}
+		return nil
+	}
+
+	switch r.Descriptor().Name {
+	case meshservice_api.MeshServiceType:
+		// find MeshService's Mesh and compute rules for it
+		if mesh := findMesh(r.GetMeta().GetMesh()); mesh != nil {
+			return rr.Compute(mesh, l)
+		}
+	case meshextenralservice_api.MeshExternalServiceType:
+		// find MeshExternalService's Mesh and compute rules for it
+		if mesh := findMesh(r.GetMeta().GetMesh()); mesh != nil {
+			return rr.Compute(mesh, l)
+		}
+	case meshhttproute_api.MeshHTTPRouteType:
+		// todo(lobkovilya): handle MeshHTTPRoute
+	}
+
+	return nil
+}
+
+func BuildResourceRules(list []PolicyItemWithMeta, l ResourceLister) (ResourceRules, error) {
+	rules := ResourceRules{}
+
+	SortByTargetRefV2(list)
+
+	var resolvedItems []*resolvedPolicyItem
+	for _, item := range list {
+		resolvedItems = append(resolvedItems, resolveTargetRef(item, l)...)
+	}
+
+	for _, i := range resolvedItems {
```

Comment:
can we change naming of this variables form i, j to something more concrete that will be easier to understand?

Reply:
I rewrote this function to be more readable, no more i,j in the code